README
License for all but the tika.jar file: GNU GPL 2.0
   tika.jar is licensed with the ASF License (Apache)

Installation: See, INSTALL.txt

Description:

eZ Tika is an extension that enables a handler for converting multiple binary file formats to plain text as used by the search engine (if you enabled those attributes as searcheable)

Currently, most common office formats are enabled (see also binaryfile.ini.append.php):

[application/pdf]
[application/msword]
[application/vnd.ms-excel]
[application/vnd.ms-powerpoint]
[application/vnd.visio]
[application/vnd.ms-outlook]
[application/xml]
[application/rtf]
[application/vnd.oasis.opendocument.text]
[application/vnd.oasis.opendocument.presentation]
[application/vnd.oasis.opendocument.spreadsheet]
[application/vnd.oasis.opendocument.formula]
[application/zip]
[application/vnd.openxmlformats-officedocument.wordprocessingml.document]
[application/vnd.openxmlformats-officedocument.spreadsheetml.sheet]
[application/vnd.openxmlformats-officedocument.presentationml.presentation]
[application/octet-stream]


CHANGELOG.md
Changelog eztika 1.15 to 2.0
----------------------------

- upgraded bundled tika.jar to 2.5.0

- added testing of the extension via GitHub Actions

Changelog eztika 1.14 to 1.15
----------------------------

- upgraded bundled tika.jar to 1.28.5

Changelog eztika 1.13 to 1.14
----------------------------

- upgraded bundled tika.jar to 1.28, fixing security issues related to log4j

Changelog eztika 1.12 to 1.13
----------------------------

- upgraded bundled tika.jar to 1.25
- improved the shell scripts used to start Tika and PdfToText: they can now be configured via usage of env vars JAVA,
  TIKA_JAR, TIKA_LOG4J and PDFTOTEXT

Changelog eztika 1.11 to 1.12
----------------------------

- upgraded bundled tika.jar to 1.24.1

Changelog eztika 1.10 to 1.11
----------------------------

- upgraded bundled tika.jar to 1.23

Changelog eztika 1.9 to 1.10
----------------------------

- upgraded bundled tika.jar to 1.22
- fixed warning "mktime(): You should be using the time() function instead"

Changelog eztika 1.8 to 1.9
---------------------------

- upgraded bundled tika.jar to 1.18

Changelog eztika 1.7 to 1.8
---------------------------

- upgraded bundled tika.jar to 1.14

Changelog eztika 1.6 to 1.7
---------------------------

- one improvement to the multiparser

Changelog eztika 1.5 to 1.6
---------------------------

- minor changes

Changelog eztika 1.4 to 1.5
---------------------------

- adding extension.xml file
- update binaryfile.ini.append.php to use eztika executable in eztika extension by default
- new ini eztika.ini with new debug setting

        [DebugSettings]
        # Debug-enabled|disabled
        # if enabled
        # - write Debug Messages to eztika.log
        #
        # Note: an error message to error.log is always written
        # if eztika can not extract any content from binaryfile
        Debug-disabled

        # KeepTempFiles-enabled|disabled
        # if enabled var/cache/ eztika_xxx.txt tmp files are not unlinked
        # to debug metadata which is extracted from the binaryfile
        # The setting is only active if Debug-enabled
        KeepTempFiles-disabled

Changelog eztika 1.3 to 1.4
---------------------------

- updated tika.jar to 1.0-dev (svn rev 1156078); adding support for chm, iWork and many bugfixes

Changelog eztika 1.2 to 1.3
---------------------------
- updated tika.jar to 0.8-dev (svn rev 933934); now supporting correctly CJK pdfs

Changelog eztika 1.1 to 1.2
---------------------------

- updated tika.jar to 0.6-dev (rev 897576) which has better support for ms excel, a reduced footprint and output encoding options
  it now correctly converts OOo and MSxx formats with asian content to UTF-8
- added encoding option --encoding-utf8 to eztika wrapper script
- updated ezinfo.php
- changed structure for building with http://projects.ez.no/ezextensionbuilder

Changelog eztika 1.0 to 1.1
---------------------------

- (added  2010-01-17) let ezpdftotext specify the no pagebreaks option, which potentially break the UTF payload sent to ez find
- updated tika.jar to 0.5-dev (rev  814142), including more support for office xml formats and various bugfixes
- added office xml format mimetypes to binaryfile.ini
- updated ezinfo.php


Tests/phpunit/01IndexationTest.php
<?php

include_once(__DIR__.'/CommandExecutingTest.php');

class CommandsTest extends CommandExecutingTest
{
    static $backupArgv;

    public function testIndexation()
    {
        $sql = "SELECT object_count FROM ezsearch_word WHERE word = 'ceci'";
        $conn = $this->getContainer()->get('ezpublish.connection')->getConnection();
        $stmt = $conn->executeQuery($sql);
        $result = $stmt->fetch();
        $before = (int)$result['object_count'];

        $this->runMigration('Tests/dsl/good/IndexationTest001.yml');

        $sql = "SELECT object_count FROM ezsearch_word WHERE word = 'ceci'";
        $conn = $this->getContainer()->get('ezpublish.connection')->getConnection();
        $stmt = $conn->executeQuery($sql);
        $result = $stmt->fetch();
        $this->assertEquals(2, (int)$result['object_count'] - $before);

        $this->runMigration('Tests/dsl/good/IndexationTest001_post.yml');
    }

    protected function runMigration($path, array $params = array())
    {
        /// @todo should we first remove the migration if it was already run?
        $params = array_merge($params, array('--path' => array($path), '-n' => true, '-u' => true));
        $out = $this->runCommand('kaliop:migration:migrate', $params);
        // check that there are no notes related to adding the migration before execution
        //$this->assertNotContains('Skipping ' . basename($path), $out, "Migration definition is incorrect?");
        $this->assertRegexp('?\| ' . basename($path) . ' +\| +\|?', $out);
        return $out;
    }
}


Tests/phpunit/CommandExecutingTest.php
<?php

use eZ\Bundle\EzPublishCoreBundle\Console\Application;
use PHPUnit\Framework\TestCase;
use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
use Symfony\Component\Console\Input\ArrayInput;
use Symfony\Component\Console\Output\StreamOutput;

abstract class CommandExecutingTestBase extends KernelTestCase
{
    protected $leftovers = array();

    /** @var \Symfony\Component\DependencyInjection\ContainerInterface $container */
    private $_container;
    /** @var \eZ\Bundle\EzPublishCoreBundle\Console\Application $app */
    protected $app;
    /** @var StreamOutput $output */
    protected $output;

    // tell to phpunit not to mess with ezpublish legacy global vars...
    protected $backupGlobalsBlacklist = array('eZCurrentAccess');

    protected static $backupArgv;

    protected function doSetUp()
    {
        $this->_container = $this->bootContainer();

        $this->app = new Application(static::$kernel);
        $this->app->setAutoExit(false);
        $fp = fopen('php://temp', 'r+');
        $this->output = new StreamOutput($fp);
        $this->leftovers = array();
    }

    /**
     * Fetches the data from the output buffer, resetting it.
     * It would be nice to use BufferedOutput, but that is not available in Sf 2.3...
     * @return null|string
     */
    protected function fetchOutput()
    {
        if (!$this->output) {
            return null;
        }

        $fp = $this->output->getStream();
        rewind($fp);
        $out = stream_get_contents($fp);

        fclose($fp);
        $fp = fopen('php://temp', 'r+');
        $this->output = new StreamOutput($fp);

        return $out;
    }

    protected function doTearDown()
    {
        foreach ($this->leftovers as $file) {
            unlink($file);
        }

        // clean buffer, just in case...
        if ($this->output) {
            $fp = $this->output->getStream();
            fclose($fp);
            $this->output = null;
        }

        // shuts down the kernel etc...
        // Since we added the BC trick with doSetup/doTeardown, this would be a loop. So we do all here
        //parent::tearDown();
        static::ensureKernelShutdown();
        static::$kernel = null;
    }

    /**
     * @return \Symfony\Component\DependencyInjection\ContainerInterface
     * @throws Exception
     */
    protected function bootContainer()
    {
        static::ensureKernelShutdown();

        if (!isset($_SERVER['SYMFONY_ENV'])) {
            throw new \Exception("Please define the environment variable SYMFONY_ENV to specify the environment to use for the tests");
        }
        // Run in our own test environment. Sf by default uses the 'test' one. We let phpunit.xml set it...
        // We also allow to disable debug mode
        $options = array(
            'environment' => $_SERVER['SYMFONY_ENV']
        );
        if (isset($_SERVER['SYMFONY_DEBUG'])) {
            $options['debug'] = $_SERVER['SYMFONY_DEBUG'];
        }
        try {
            static::bootKernel($options);
        } catch (\RuntimeException $e) {
            throw new \RuntimeException($e->getMessage() . " Did you forget to define the environment variable KERNEL_DIR?", $e->getCode(), $e->getPrevious());
        }

        // In Sf4 we do have the container available, in Sf3 we do not
        return isset(static::$container) ? static::$container : static::$kernel->getContainer();
    }

    protected function getContainer()
    {
        return $this->_container;
    }

    /**
     * @param string $commandName
     * @param array $params
     * @param bool $checkExitCode
     * @return string|null
     * @throws Exception
     */
    protected function runCommand($commandName, array $params = array(), $checkExitCode = true)
    {
        $before = getcwd();
        $exitCode = $this->app->run($this->buildInput($commandName, $params), $this->output);
        $output = $this->fetchOutput();
        if ($checkExitCode) {
            $this->assertSame(0, $exitCode, 'CLI Command failed. Output: ' . $output);
        }
        // in case legacy kernel messes things up
        chdir($before);
        return $output;
    }

    /**
     * @param string $legacyScript legacy script
     * @param array $args used to inject arguments and options into legacy cli scripts
     * @return string
     */
    public function runLegacyScript($legacyScript, $args = array())
    {
        // There is a bug in LegacyEmbedScriptCommand when running many legacy scripts in a row. We hack around it

        $container = $this->getContainer();

        if (!is_array(self::$backupArgv)) {
            self::$backupArgv = $GLOBALS['argv'];

            // first, we make sure not to try to instantiate the web-kernel-handler
            $ch = $this->getContainer()->get('ezpublish_legacy.kernel_handler.cli');
            $container->set( 'ezpublish_legacy.kernel.lazy', null );
            $container->set( 'ezpublish_legacy.kernel_handler', $ch );
            $container->set( 'ezpublish_legacy.kernel_handler.web', $ch );
        }

        // then we get the kernel
        /** @var callable $k */
        $k = $container->get('ezpublish_legacy.kernel');
        $k = $k();
        // and patch *its own* cli handler
        /** @var \eZ\Publish\Core\MVC\Legacy\Kernel\CLIHandler $cl */
        $ch = \Closure::bind(function () { return $this->kernelHandler; }, $k, 'ezpKernel');
        $ch = $ch();
        $ch->setEmbeddedScriptPath($legacyScript);

        // This is required to pass in legacy-script arguments
        $GLOBALS['argv'] = array_merge(self::$backupArgv, array($legacyScript), $args);
        $_SERVER['argv'] = $GLOBALS['argv'];

        // CLIHandler does not close its output buffering. We do it, or phpunit will mark the test as risky.
        // We also do our own buffering to avoid polluting test output
        ob_start();
        $this->runCommand('ezpublish:legacy:script', array('script' => $legacyScript));
        ob_get_contents();
        ob_end_clean();
        $output = ob_get_contents();
        ob_end_clean();

        return $output;
    }

    /**
     * @param $commandName
     * @param array $params
     * @return ArrayInput
     */
    protected function buildInput($commandName, array $params = array())
    {
        $params = array_merge(['command' => $commandName], $params);
        return new ArrayInput($params);
    }
}

// Auto-adapt to PHPUnit 8 that added a `void` return-type to the setUp/tearDown methods
/// @todo check: can we leave this to the parent class from Symfony?
if (method_exists(\ReflectionMethod::class, 'hasReturnType') && (new \ReflectionMethod(TestCase::class, 'tearDown'))->hasReturnType()) {
    // eval is required for php 5.6 compatibility
    eval('abstract class CommandExecutingTest extends CommandExecutingTestBase
    {
        protected function setUp(): void
        {
            $this->doSetUp();
        }

        protected function tearDown(): void
        {
            $this->doTearDown();
        }
    }');
}else {
    abstract class CommandExecutingTest extends CommandExecutingTestBase
    {
        protected function setUp()
        {
            $this->doSetUp();
        }

        protected function tearDown()
        {
            $this->doTearDown();
        }
    }
}


ezinfo.php
<?php
//
// ## BEGIN COPYRIGHT, LICENSE AND WARRANTY NOTICE ##
// SOFTWARE NAME: eZ Tika
// SOFTWARE RELEASE: 1.4-0-final
// COPYRIGHT NOTICE: Copyright (C) 2008-2009 Paul Borgermans
// SOFTWARE LICENSE: GNU General Public License v2.0
// NOTICE: >
//   This program is free software; you can redistribute it and/or
//   modify it under the terms of version 2.0  of the GNU General
//   Public License as published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of version 2.0 of the GNU General
//   Public License along with this program; if not, write to the Free
//   Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
//   MA 02110-1301, USA.
//
//
// ## END COPYRIGHT, LICENSE AND WARRANTY NOTICE ##
//

/**
 * Display meta and copyright info for eZ Publish eztika extension
 */
class eZTikaInfo
{
    static function info()
    {
        return array(
            'Name' => "eZ Tika",
            'Version' => "2.0.0",
            'Copyright' => "Copyright 2008-2022 Paul Borgermans",
            'Info_url' => "https://github.com/xrowgmbh/eztika",
            'License' => "GNU General Public License v2.0",
            '3rdparty_software' =>
                        array ( 'name' => 'Tika',
                                'Version' => '2.5.0',
                                'copyright' => 'The Apache Software Foundation.',
                                'license' => 'Apache License, Version 2.0',
                                'info_url' => 'https://tika.apache.org/' ) );
        }
}


settings/binaryfile.ini.append.php
<?php /*

[HandlerSettings]
ExtensionRepositories[]=eztika/classes/datatypes/ezbinaryfile

MetaDataExtractor[application/pdf]=ezmulti
MetaDataExtractor[application/msword]=ezmulti
MetaDataExtractor[application/vnd.ms-excel]=ezmulti
MetaDataExtractor[application/vnd.ms-powerpoint]=ezmulti
MetaDataExtractor[application/vnd.visio]=ezmulti
MetaDataExtractor[application/vnd.ms-outlook]=ezmulti
MetaDataExtractor[application/xml]=ezmulti
MetaDataExtractor[application/rtf]=ezmulti
MetaDataExtractor[application/vnd.oasis.opendocument.text]=ezmulti
MetaDataExtractor[application/vnd.oasis.opendocument.presentation]=ezmulti
MetaDataExtractor[application/vnd.oasis.opendocument.spreadsheet]=ezmulti
MetaDataExtractor[application/vnd.oasis.opendocument.formula]=ezmulti
MetaDataExtractor[application/zip]=ezmulti
# Office OOXML files, supported since eZP4.0.4, eZP4.1.0
MetaDataExtractor[application/vnd.openxmlformats-officedocument.wordprocessingml.document]=ezmulti
MetaDataExtractor[application/vnd.openxmlformats-officedocument.spreadsheetml.sheet]=ezmulti
MetaDataExtractor[application/vnd.openxmlformats-officedocument.presentationml.presentation]=ezmulti
#octet-stream is the default
# If tika senses a known format like the Office 2007 ooxml types, it will treat them well
MetaDataExtractor[application/octet-stream]=ezmulti

[MultiHandlerSettings]
# Change the path to the eztika shell script, which probably needs editing as well
TextExtractionTool=./extension/eztika/bin/eztika


settings/binaryfile.ini.append.withxpdf.php
<?php /*

[HandlerSettings]
ExtensionRepositories[]=eztika/classes/datatypes/ezbinaryfile

# warning: CJK pdf's are not working correctly, use xpdf based conversion instead
MetaDataExtractor[application/pdf]=ezpdf

MetaDataExtractor[application/msword]=ezmulti
MetaDataExtractor[application/vnd.ms-excel]=ezmulti
MetaDataExtractor[application/vnd.ms-powerpoint]=ezmulti
MetaDataExtractor[application/vnd.visio]=ezmulti
MetaDataExtractor[application/vnd.ms-outlook]=ezmulti
MetaDataExtractor[application/xml]=ezmulti
MetaDataExtractor[application/rtf]=ezmulti
MetaDataExtractor[application/vnd.oasis.opendocument.text]=ezmulti
MetaDataExtractor[application/vnd.oasis.opendocument.presentation]=ezmulti
MetaDataExtractor[application/vnd.oasis.opendocument.spreadsheet]=ezmulti
MetaDataExtractor[application/vnd.oasis.opendocument.formula]=ezmulti
MetaDataExtractor[application/zip]=ezmulti
# Office OOXML files, supported since eZP4.0.4, eZP4.1.0
MetaDataExtractor[application/vnd.openxmlformats-officedocument.wordprocessingml.document]=ezmulti
MetaDataExtractor[application/vnd.openxmlformats-officedocument.spreadsheetml.sheet]=ezmulti
MetaDataExtractor[application/vnd.openxmlformats-officedocument.presentationml.presentation]=ezmulti
#octet-stream is the default
# If tika senses a known format like the Office 2007 ooxml types, it will treat them well
MetaDataExtractor[application/octet-stream]=ezmulti

[MultiHandlerSettings]

# Change the path to the eztika shell script, which probably needs editing as well
TextExtractionTool=./extension/eztika/bin/eztika

[PDFHandlerSettings]
TextExtractionTool=./extension/eztika/bin/ezpdftotext


classes/datatypes/ezbinaryfile/plugins/ezmultiparser.php
<?php
//
// Definition of eZMultiParser class
//
// Created on: <08-Dec-2008 20:13 pb>
//
// ## BEGIN COPYRIGHT, LICENSE AND WARRANTY NOTICE ##
// SOFTWARE NAME: eZ Tika
// SOFTWARE RELEASE: 1.4-0-final
// COPYRIGHT NOTICE: Copyright (C) 2008-2009 Paul Borgermans
// SOFTWARE LICENSE: GNU General Public License v2.0
// NOTICE: >
//   This program is free software; you can redistribute it and/or
//   modify it under the terms of version 2.0  of the GNU General
//   Public License as published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of version 2.0 of the GNU General
//   Public License along with this program; if not, write to the Free
//   Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
//   MA 02110-1301, USA.
//
//
// ## END COPYRIGHT, LICENSE AND WARRANTY NOTICE ##
//

/**
 * Handles parsing of multiple files and returns the metadata
 */
class eZMultiParser
{
    // if true debug messages will be write to eztika.log
    protected $DebugIsEnabled = false;

    // if true the tmp cache file for extracted text is not unlinked
    protected $DebugKeepTempFiles = false;

    function __construct()
    {
        $eztikaINI = eZINI::instance( 'eztika.ini' );
        $debugEnabled = $eztikaINI->variable( 'DebugSettings', 'Debug' );
        if ( $debugEnabled == 'enabled' )
        {
            $this->DebugIsEnabled = true;
            $keepTempFiles = $eztikaINI->variable( 'DebugSettings', 'KeepTempFiles' );
            if ( $keepTempFiles == 'enabled' )
            {
                $this->DebugKeepTempFiles = true;
            }

        }
    }

    function parseFile( $fileName )
    {
        if( !$fileName )
        {
            eZDebug::writeError( "eztika cannot find the file $fileName; skipping", 'eztika class eZMultiParser' );
            return false;
        }

        $originalFileSize = filesize( $fileName );
        $this->writeEzTikaLog( '[START] eZMultiParser for File: '. round( $originalFileSize/1024, 2 ) .' KByte '. $fileName );

        $binaryINI = eZINI::instance( 'binaryfile.ini' );
        $textExtractionTool = $binaryINI->variable( 'MultiHandlerSettings', 'TextExtractionTool' );

        $startTime = time();

        $tmpName = eZSys::cacheDirectory() . eZSys::fileSeparator(). 'eztika_'. md5( $startTime ) . '.txt';
        $handle = fopen( $tmpName, "w" );
        fclose( $handle );
        chmod( $tmpName, 0777 );

        $cmd = "$textExtractionTool $fileName > $tmpName";

        $this->writeEzTikaLog( 'exec: ' . $cmd );

        // perform eztika command
        exec( $cmd , $returnArray, $returnCode );

        $this->writeEzTikaLog( "exec returnCode: $returnCode" );

        $metaData = '';
        if ( file_exists( $tmpName ) )
        {
            $fp = fopen( $tmpName, 'r' );
            $fileSize = filesize( $tmpName );
            $metaData = fread( $fp, $fileSize );
            fclose( $fp );

            // keep tempfile in debugmode
            if ( $this->DebugKeepTempFiles === true )
            {
                $this->writeEzTikaLog( 'keep tempfile for debugging extracted metadata: '. $tmpName );
            }
            else
            {
                $this->writeEzTikaLog( 'unlink tempfile: '. $tmpName );
                unlink( $tmpName );
            }

            if ( $fileSize === false || $fileSize === 0 )
            {
                $this->writeEzTikaLog( "[ERROR] no metadata was extracted! Check if eztika is working correctly" );
            }
            else
            {
                $this->writeEzTikaLog( 'metadata read from tempfile '. round( $fileSize / 1024, 2 ) . ' KByte' );
            }
        }
        else
        {
            $this->writeEzTikaLog( "[ERROR] no tempfile '$tmpName' for eztika output exists,
                check if eztika command is working,
                check if eztika is executeable" );
        }

        // write an error message to error.log if no data could be extracted
        if ( $metaData == '' && $originalFileSize > 0 )
        {
            eZDebug::writeError( "eztika can not extract content from binaryfile for searchindex \nexec( $cmd )", 'eztika class eZMultiParser' );
        }

        $endTime = time();
        $seconds = $endTime - $startTime;
        $this->writeEzTikaLog( "[END] after $seconds s" );

        return $metaData;
    }

    /**
     *
     * write log message to eztika.log
     * @param unknown_type $message
     */
    private function writeEzTikaLog( $message )
    {
        if ( $this->DebugIsEnabled )
            eZLog::write( $message, 'eztika.log', $this->logDirectory() );
    }

    /**
     * @return the directory used for storing log files.
     */
    private function logDirectory( )
    {
        $ini = eZINI::instance( );
        $logDir = $ini->variable( 'FileSettings', 'LogDir' );

        if ( $logDir[0] == "/" )
        {
            return eZDir::path( array( $logDir ) );
        }
        else
        {
            return eZDir::path( array( eZSys::varDirectory(), $logDir ) );
        }
    }
}
